File: src/vibe_kanban_clone/api/routes/agents.py
"""Agent routes."""

from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from vibe_kanban_clone.api.deps import get_session
from vibe_kanban_clone.schemas.agent import AgentCreate, AgentRead, AgentUpdate
from vibe_kanban_clone.schemas.common import PaginatedParams
from vibe_kanban_clone.services import agents as agents_service
from vibe_kanban_clone.services import skills as skills_service

router = APIRouter()


@router.get("/agents", response_model=list[AgentRead])
async def list_agents(
    pagination: Annotated[PaginatedParams, Depends()],
    session: Annotated[AsyncSession, Depends(get_session)],
) -> list[AgentRead]:
    """List all agents."""
    return await agents_service.list_agents(session, pagination.limit, pagination.offset)


@router.post("/agents", response_model=AgentRead, status_code=201)
async def create_agent(
    data: AgentCreate,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> AgentRead:
    """Create a new agent."""
    agent = await agents_service.create_agent(session, data)
    return agent


@router.get("/agents/{agent_id}", response_model=AgentRead)
async def get_agent(
    agent_id: int,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> AgentRead:
    """Get an agent by ID."""
    agent = await agents_service.get_agent(session, agent_id)
    if agent is None:
        raise HTTPException(status_code=404, detail="Agent not found")
    return agent


@router.patch("/agents/{agent_id}", response_model=AgentRead)
async def update_agent(
    agent_id: int,
    data: AgentUpdate,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> AgentRead:
    """Update an agent."""
    agent = await agents_service.get_agent(session, agent_id)
    if agent is None:
        raise HTTPException(status_code=404, detail="Agent not found")
    agent = await agents_service.update_agent(session, agent, data)
    return agent


@router.delete("/agents/{agent_id}", status_code=204)
async def delete_agent(
    agent_id: int,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> None:
    """Delete an agent."""
    agent = await agents_service.get_agent(session, agent_id)
    if agent is None:
        raise HTTPException(status_code=404, detail="Agent not found")
    await agents_service.delete_agent(session, agent)


@router.post("/agents/{agent_id}/skills/{skill_id}", response_model=AgentRead)
async def assign_skill(
    agent_id: int,
    skill_id: int,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> AgentRead:
    """Assign a skill to an agent."""
    agent = await agents_service.get_agent(session, agent_id)
    if agent is None:
        raise HTTPException(status_code=404, detail="Agent not found")
    skill = await skills_service.get_skill(session, skill_id)
    if skill is None:
        raise HTTPException(status_code=404, detail="Skill not found")
    agent = await agents_service.assign_skill(session, agent, skill)
    return agent


@router.delete("/agents/{agent_id}/skills/{skill_id}", status_code=204)
async def remove_skill(
    agent_id: int,
    skill_id: int,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> None:
    """Remove a skill from an agent."""
    agent = await agents_service.get_agent(session, agent_id)
    if agent is None:
        raise HTTPException(status_code=404, detail="Agent not found")
    skill = await skills_service.get_skill(session, skill_id)
    if skill is None:
        raise HTTPException(status_code=404, detail="Skill not found")
    await agents_service.remove_skill(session, agent, skill)

File: server/src/routes/agents.ts
/**
 * Routes: /api/agents
 *   GET    /agents                - list all agents (paginated)
 *   POST   /agents                - create a new agent
 *   GET    /agents/:id            - get agent detail (with skills)
 *   PATCH  /agents/:id            - update agent
 *   DELETE /agents/:id            - delete agent
 *   POST   /agents/:id/skills/:skillId - assign skill to agent
 *   DELETE /agents/:id/skills/:skillId - remove skill from agent
 *   POST   /agents/:id/tools/:toolId   - assign tool to agent
 *   DELETE /agents/:id/tools/:toolId   - remove tool from agent
 */
import { zValidator } from "@hono/zod-validator";
import { Hono } from "hono";
import { HTTPException } from "hono/http-exception";
import { AgentCreate, AgentUpdate } from "../schemas/agent.js";
import { PaginatedParams } from "../schemas/common.js";
import * as agentsService from "../services/agents.js";
import * as skillsService from "../services/skills.js";
import * as toolsService from "../services/tools.js";
import { broadcastGlobal } from "../ws/manager.js";

const app = new Hono();

app.get("/agents", zValidator("query", PaginatedParams), async (c) => {
  const { limit, offset } = c.req.valid("query");
  const agents = await agentsService.listAgents(limit, offset);
  return c.json(agents);
});

app.post("/agents", zValidator("json", AgentCreate), async (c) => {
  const data = c.req.valid("json");
  const agent = await agentsService.createAgent(data);
  await broadcastGlobal("agent.created", agent);
  return c.json(agent, 201);
});

app.get("/agents/:id", async (c) => {
  const id = Number(c.req.param("id"));
  const agent = await agentsService.getAgent(id);
  if (!agent) throw new HTTPException(404, { message: "Agent not found" });
  return c.json(agent);
});

app.patch("/agents/:id", zValidator("json", AgentUpdate), async (c) => {
  const id = Number(c.req.param("id"));
  const agent = await agentsService.getAgent(id);
  if (!agent) throw new HTTPException(404, { message: "Agent not found" });
  const data = c.req.valid("json");
  const updated = await agentsService.updateAgent(id, data);
  await broadcastGlobal("agent.updated", updated);
  return c.json(updated);
});

app.delete("/agents/:id", async (c) => {
  const id = Number(c.req.param("id"));
  const agent = await agentsService.getAgent(id);
  if (!agent) throw new HTTPException(404, { message: "Agent not found" });
  await agentsService.deleteAgent(id);
  await broadcastGlobal("agent.deleted", { agent_id: id });
  return c.body(null, 204);
});

app.post("/agents/:id/skills/:skillId", async (c) => {
  const id = Number(c.req.param("id"));
  const skillId = Number(c.req.param("skillId"));
  const agent = await agentsService.getAgent(id);
  if (!agent) throw new HTTPException(404, { message: "Agent not found" });
  const skill = await skillsService.getSkill(skillId);
  if (!skill) throw new HTTPException(404, { message: "Skill not found" });
  const updated = await agentsService.assignSkill(id, skillId);
  await broadcastGlobal("agent.updated", updated);
  return c.json(updated);
});

app.delete("/agents/:id/skills/:skillId", async (c) => {
  const id = Number(c.req.param("id"));
  const skillId = Number(c.req.param("skillId"));
  const agent = await agentsService.getAgent(id);
  if (!agent) throw new HTTPException(404, { message: "Agent not found" });
  const skill = await skillsService.getSkill(skillId);
  if (!skill) throw new HTTPException(404, { message: "Skill not found" });
  const updated = await agentsService.removeSkill(id, skillId);
  await broadcastGlobal("agent.updated", updated);
  return c.json(updated);
});

app.post("/agents/:id/tools/:toolId", async (c) => {
  const id = Number(c.req.param("id"));
  const toolId = Number(c.req.param("toolId"));
  const agent = await agentsService.getAgent(id);
  if (!agent) throw new HTTPException(404, { message: "Agent not found" });
  const tool = await toolsService.getTool(toolId);
  if (!tool) throw new HTTPException(404, { message: "Tool not found" });
  const updated = await agentsService.assignTool(id, toolId);
  await broadcastGlobal("agent.updated", updated);
  return c.json(updated);
});

app.delete("/agents/:id/tools/:toolId", async (c) => {
  const id = Number(c.req.param("id"));
  const toolId = Number(c.req.param("toolId"));
  const agent = await agentsService.getAgent(id);
  if (!agent) throw new HTTPException(404, { message: "Agent not found" });
  const tool = await toolsService.getTool(toolId);
  if (!tool) throw new HTTPException(404, { message: "Tool not found" });
  const updated = await agentsService.removeTool(id, toolId);
  await broadcastGlobal("agent.updated", updated);
  return c.json(updated);
});

export default app;
